ledger/simulation/tracer.go

```diff
@@ -130,8 +130,14 @@ func (tracer *evalTracer) BeforeTxnGroup(ep *logic.EvalParams) {
 	if ep.GetCaller() != nil {
 		// If this is an inner txn group, save the txns
 		tracer.populateInnerTransactions(ep.TxnGroup)
+		tracer.result.TxnGroups[0].BudgetAdded += uint64(ep.Proto.MaxAppProgramCost)
 	}
 	tracer.cursorEvalTracer.BeforeTxnGroup(ep)
+
+	// Currently only supports one (first) txn group
+	if ep.PooledApplicationBudget != nil && tracer.result.TxnGroups[0].BudgetAdded == 0 {
```

Comment:
We initialize the group budget by only accounting for top-level transactions. If we are in an inner level txn, then we dynamically add `ep.Proto.MaxAppProgramCost` to the budget each time. We might also want to think about how this can be handled when an app has depleted all its funds before it tries to "op-up" in an inner transaction and whether simulation should report an error in this case (like real execution), or simulate anyway.

Reply:
> when an app has depleted all its funds before it tries to "op-up" in an inner transaction and whether simulation should report an error in this case

That is what we discussed yesterday about keeping track of budget as execution goes, right?

Reply:
I can understand that, the second if seems mutual exclusive with the first if, for first if condition makes `BudgetAdded != 0`, thus second if cannot be entered.

Could you explain more on what scenario the first if and second if will be triggered, such that it won't seem too abstract.

Reply:
I forgot to clarify this in the previous comment:
* I considered a case where `ep.PooledApplicationBudget` might  be nil because of an old pragma version (before pooling was enabled, this pointer would just be nil). Maybe for completion's sake, we should specify that the budget = `ep.Proto.MaxAppProgramCost` if this value is nil (since you cannot pool), and if it's not nil, we start adding the budget.
* The `BudgetAdded == 0` checks that we are always at the top-level transaction. If we are at an inner transaction, our `PooledApplicationBudget` could be something like `567` after being called midway through the app call. So we are saying that at the root-level transaction, set the `BudgetAdded` to a value that only accounts for the top level txns (e.g. 2100 for 3 txns) and then in the inner txn, we only add 700 once.

Reply:
I think it's ok to do nothing here if `EnableAppCostPooling` is false (i.e. the current code looks good to me).

If I understand the code correctly, `BudgetAdded` for the group would then be 0 and get omitted from the response. This makes sense, since if app budgets aren't pooled, there's not much use in aggregating the budgets for the whole group anyway. And of course individual transactions' `BudgetUsed` would still be accurate in that case.

Reply:
thanks for explaining! that helps me a lot in understanding and catching up!